src/nvim/ui_client.c

```diff
@@ -283,6 +284,60 @@ void ui_client_event_raw_line(GridLineEvent *g)
                (const schar_T *)grid_line_buf_char, grid_line_buf_attr);
 }
 
+void ui_client_event_connect(Array args)
+{
+  if (args.size < 2 || args.items[0].type != kObjectTypeString || args.items[1].type != kObjectTypeBoolean) {
+    ELOG("Error handling UI event 'connect'");
+    return;
+  }
+
+  char *server_addr = args.items[0].data.string.data;
+  bool stop_server = args.items[1].data.boolean;
+
+  ELOG("stop_server: %d", stop_server);
+  multiqueue_put(main_loop.fast_events, channel_connect_event, server_addr, (void *) stop_server);
```

Comment:
just for troubleshooting, something to try:
```suggestion
  multiqueue_put(main_loop.events, channel_connect_event, server_addr, (void *) stop_server);
```

Reply:
Putting the event in `main_loop.events` like this instead of `fast_events` just delays the execution of the `channel_connect_event`. No connection is created, instead the same assertion failure (`!ui_client_channel_id`) happens when you try exiting neovim instead of it happening immediately.